moby/moby
# IPv6 only: use random MAC addresses

**- What I did**

In a bridge network with IPv4 enabled, when no MAC address is explicitly configured for a container, the MAC address was generated from the IPv4 address. So, when the container was stopped and its IP addresses were quickly recycled for another container, ARP/neighbour caches were still correct because there was a fixed relationship between MAC and IP addresses.

It's not possible to generate a unique MAC address from an IPv6 address (at least, not with a prefix length of fewer than 80 bits, and the prefix length should normally be at-least 64-bits).

So, for a container in an IPv6-only network, a new scheme is needed or when an IPv6 address gets recycled for another container the neighbour cache will contain the wrong mapping from that IPv6 address to a MAC address.

- Also, fixes https://github.com/moby/moby/issues/48082
  - An issue with Portainer setting a MAC address but no IP address when restarting containers, resulting in duplicate MAC addresses derived from recycled IP addresses. Now, the old random MAC address won't be recycled along with the IP address.

**- How I did it**

- Abandoned the fixed mapping from IPv4 address to MAC address.
  - Now, container interfaces on a bridge network always have a randomly generated MAC address.
  - Send an unsolicited ARP or Neighbour Advertisement when bringing up a new interface, to update caches.

Some of this was written as prototype by @corhere, I've picked up some changes from [this branch](https://github.com/moby/moby/compare/master...corhere:moby:libn/gratuitous-arp). However, setting sysctl `arp_notify` to have the kernel send the unsolicited ARP/NA messages didn't work out ...
- the kernel didn't send an ARP/NA when the interface came up
  - https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1989187
- as a workaround, we tried triggering the ARP/NA by setting the interface's MAC address
  - that was fine for a single initial message, but
  - setting the MAC address again later to trigger ARP/NA resends, flushed entries from the ARP/neighbour caches.

So, the ARP/NA messages are now sent from user space.

The number of unsolicited ARP/NA messages, and the interval between them, are configurable:
- The default is to send 3 messages at 1s intervals.
- That can be overridden in "docker network create" using:
    `-o com.docker.network.advertise_addr_nmsgs=3`
    `-o com.docker.network.advertise_addr_ms=1000`
- Or, in daemon.json for each driver:
    ```
        "default-network-opts": {
          "bridge": {
            "com.docker.network.advertise_addr_nmsgs": "3",
            "com.docker.network.advertise_addr_ms": "1000"
          }
        }
    ```
- The allowed range is 0-3 for count, and 100-2000ms for the interval.
- Setting count to 0 disables the unsolicited ARP/NA messages.
- The default bridge will always use the built-in defaults, it cannot be configured.

Note that random MAC addresses are now used on the default bridge network.
- Until now, as a special case for the default bridge network, IPv6 addresses have been based on the MAC address rather than being IPAM-assigned. So, as the MAC address was based on the IPv4 address unless it was specifically configured, there was a fixed mapping from IPv4 to MAC to IPv6 address. That no longer holds. But, all three addresses can be explicitly configured.

**- How to verify it**

New integration test to check that ARP/NA messages are sent.

Also tested while running `iperf` ...
- In a moby dev container, running in DD on an M2 Macbook.
- `iperf -s` in one container on a bridge, and `iperf -c <addr>` on another -> ~115Gbps TCP
- In one of the containers, pinged a third container's v4 and v6 addresses, to populate the neighbour cache.
  - Restarted the third container a few times, with the same IP address.
  - Used `ip neigh show` to check the iperf container's neighbour cache was immediately updated each time.
- Repeated that, but pinging from the other iperf container.
- Just to be sure, checked that with `nmsgs=0`, the neighbour cache wasn't updated when the third container restarted.

To check for goroutine leaks ...
- Started a bunch of containers running a 4s sleep (so that all three ARP and NA messages would be sent)...
  - `i=0; while [ $i -lt 500 ]; do ( docker run --rm -d --network b46 busybox sleep 4 & ); i=$((i+1)); done`
- Then the same again, but with a 1s sleep so that the ARP/NA goroutine would be cancelled.
- `docker info` reported that it got back to 49 goroutines.

**- Description for the changelog**
```markdown changelog
- Container interfaces in bridge networks now use randomly generated MAC addresses.
  - Unsolicited ARP / Neighbour Advertisement messages will be sent when the interfaces are started
    so that, when IP addresses are reused, they're associated with the newly generated MAC address.
  - IPv6 addresses in the default bridge network are now IPAM-assigned, rather than being derived
    from the MAC address.
  - Driver option "com.docker.network.advertise_addr_nmsgs" can be used to set the number of
    unsolicited ARP and NA packets sent for each address, from 0 to 3, the default is 3.
  - Driver option "com.docker.network.advertise_addr_ms" can be used to set the interval between
    unsolicited ARP and NA packets sent for each address, from 100 to 2000ms, the default is 1000ms.
```
